Deploying dependent document models from the same bundle are not loaded in the correct order

<!--
Provide a general summary of the issue in the Title above

If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.
-->
## I'm submitting a ...   (check one with "x")
```
[x] bug report
[ ] feature request
[ ] question
```

## Expected Behavior

I create a dynamic extension jar with 2 document models in `/META-INF/alfresco/models/`, named `XYZModel.xml` and `XYZWorkflowModel.xml`.
The `XYZWorkflowModel.xml` imports a namespace defined in `XYZModel.xml`.


I expect that both models are registered and activated when I upload the jar in the dynamic extensions control panel.

<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->
<!--- If you have a question, you can delete this -->

## Current Behavior

An exception is thrown when uploading the jar.
```
[ALFRESCO] 2018-10-29 10:30:36,839  WARN  [repo.dictionary.DictionaryRepositoryBootstrap] asynchronouslyRefreshedCacheThreadPool1] Failed to load model 'matwf:materialsWorkflowModel' : org.alfresco.service.cmr.dictionary.DictionaryException: 09290013 Failed to compile model 'matwf:materialsWorkflowModel'
 [ALFRESCO] 2018-10-29 10:30:37,001  WARN  [repo.dictionary.DictionaryRepositoryBootstrap] [asynchronouslyRefreshedCacheThreadPool1] Failed to load model 'matwf:materialsWorkflowModel' : org.alfresco.service.cmr.dictionary.DictionaryException: 09290014 Failed to compile model 'matwf:materialsWorkflowModel'
 [ALFRESCO] 2018-10-29 10:31:03,321  ERROR [extensions.webscripts.AbstractRuntime] [http-bio-8080-exec-2] Exception from executeScript - redirecting to status template error: 09290017 Failed to compile model 'matwf:materialsWorkflowModel'
 org.alfresco.service.cmr.dictionary.DictionaryException: 09290017 Failed to compile model 'xyzwf:XYZWorkflowModel'
	at org.alfresco.repo.dictionary.CompiledModel.<init>(CompiledModel.java:126)
	at org.alfresco.repo.dictionary.M2Model.compile(M2Model.java:174)
	at org.alfresco.repo.dictionary.DictionaryModelType$DictionaryModelTypeTransactionListener$2.doWork(DictionaryModelType.java:627)
	at org.alfresco.repo.security.authentication.AuthenticationUtil.runAs(AuthenticationUtil.java:555)
	at org.alfresco.repo.dictionary.DictionaryModelType$DictionaryModelTypeTransactionListener.beforeCommit(DictionaryModelType.java:574)
	at org.alfresco.util.transaction.TransactionSupportUtil$TransactionSynchronizationImpl.doBeforeCommit(TransactionSupportUtil.java:535)
	at org.alfresco.util.transaction.TransactionSupportUtil$TransactionSynchronizationImpl.doBeforeCommit(TransactionSupportUtil.java:514)
	at org.alfresco.util.transaction.TransactionSupportUtil$TransactionSynchronizationImpl.beforeCommit(TransactionSupportUtil.java:479)
	at org.springframework.transaction.support.TransactionSynchronizationUtils.triggerBeforeCommit(TransactionSynchronizationUtils.java:95)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.triggerBeforeCommit(AbstractPlatformTransactionManager.java:925)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:738)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:724)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:475)
	at org.alfresco.util.transaction.SpringAwareUserTransaction.commit(SpringAwareUserTransaction.java:482)
	at org.alfresco.repo.transaction.RetryingTransactionHelper.doInTransaction(RetryingTransactionHelper.java:486)
	at org.alfresco.repo.web.scripts.RepositoryContainer.transactionedExecute(RepositoryContainer.java:587)
	at org.alfresco.repo.web.scripts.RepositoryContainer.transactionedExecuteAs(RepositoryContainer.java:656)
	at org.alfresco.repo.web.scripts.RepositoryContainer.executeScriptInternal(RepositoryContainer.java:428)
	at org.alfresco.repo.web.scripts.RepositoryContainer.executeScript(RepositoryContainer.java:308)
	at org.springframework.extensions.webscripts.AbstractRuntime.executeScript(AbstractRuntime.java:382)
	at org.springframework.extensions.webscripts.AbstractRuntime.executeScript(AbstractRuntime.java:210)
	at org.springframework.extensions.webscripts.servlet.WebScriptServlet.service(WebScriptServlet.java:132)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:731)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.alfresco.web.app.servlet.GlobalLocalizationFilter.doFilter(GlobalLocalizationFilter.java:68)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:505)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:169)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
	at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:683)
	at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:956)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:436)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1078)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:625)
	at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:318)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.alfresco.service.namespace.NamespaceException: URI http://www.xyz.com/model/xyz/1.0 cannot be imported as it is not defined (with prefix xyz
	at org.alfresco.repo.dictionary.CompiledModel.createLocalPrefixResolver(CompiledModel.java:224)
	at org.alfresco.repo.dictionary.CompiledModel.constructDefinitions(CompiledModel.java:148)
	at org.alfresco.repo.dictionary.CompiledModel.<init>(CompiledModel.java:106)
	... 46 more
```
Reuploading the jar does not result in exceptions.
Both models are placed in the `Data Dictionary/Models` folder and the *Model Active* property is set for both models. However, the `XYZWorkflowsModel.xml` model is not fully registered.


<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->
<!--- If you have a question, you can delete this -->

## Possible Solution

The listing of the jar file includes:
```
META-INF/alfresco/
META-INF/alfresco/models/
META-INF/alfresco/models/XYZWorkflowModel.xml
META-INF/alfresco/models/XYZModel.xml
```
Note that the files are sorted in reverse alphabetical order, and I think they are read in the order they are present in the jar file.

I think dynamic extensions should catch the exception when setting the *Model Active* property and has to revert the property to `false` to accurately reflect the state of the active models.

An other possibility is to have dynamic extensions resolve the dependencies between models and activate models in the right order.

<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas how to implement the addition or change -->
<!--- If you have a question, you can delete this -->

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
<!--- If you have a question, you can delete this -->
1. Clone https://github.com/vierbergenlars/de-repro-197
2. Run `./gradlew jar` to build the dynamic extensions jar
3. Start up an alfresco with dynamic extensions
4. Upload the jar file generated in `build/libs/de-repro-197.jar` to the dynamic extensions control panel.
5. Observe exception in the alfresco logs

## Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Alfresco version used: 5.1.1 community
* DE version used: 1.7.6

--- a/alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/models/RepositoryModelRegistrar.java
+++ b/alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/models/RepositoryModelRegistrar.java
@@ -52,7 +52,7 @@ public class RepositoryModelRegistrar extends AbstractModelRegistrar {
                                 modelResource.getName(), e);
                     }
                     return null;
-                },false,false);
+                },false,true);
                 } catch(Exception e){
                     logger.error("tx error", e);
                 }

--- a/alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/models/RepositoryModelRegistrar.java
+++ b/alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/models/RepositoryModelRegistrar.java
@@ -4,7 +4,6 @@ package com.github.dynamicextensionsalfresco.models;
 import com.github.dynamicextensionsalfresco.resources.ResourceHelper;
 import org.alfresco.repo.dictionary.RepositoryLocation;
 import org.alfresco.repo.security.authentication.AuthenticationUtil;
-import org.alfresco.repo.transaction.RetryingTransactionHelper;
 import org.alfresco.service.cmr.admin.RepoAdminService;
 import org.alfresco.service.cmr.repository.NodeService;
 import org.alfresco.service.transaction.TransactionService;
@@ -13,9 +12,12 @@ import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
 
 /**
- * Register models in the repository as cm:dictionaryModel's using the [RepoAdminService].
+ * Register models in the repository as cm:dictionaryModel's using the {@link RepoAdminService}.
  */
 public class RepositoryModelRegistrar extends AbstractModelRegistrar {
+
+    private Logger logger = LoggerFactory.getLogger(RepositoryModelRegistrar.class);
+
     @Autowired
     public RepositoryLocation customModelsRepositoryLocation;
     @Autowired
@@ -26,7 +28,6 @@ public class RepositoryModelRegistrar extends AbstractModelRegistrar {
     public TransactionService transactionService;
     @Autowired
     public ResourceHelper resourceHelper;
-    private Logger logger = LoggerFactory.getLogger(RepositoryModelRegistrar.class);
 
     @Override
     public void unregisterModels() {
@@ -35,32 +36,27 @@ public class RepositoryModelRegistrar extends AbstractModelRegistrar {
 
     @Override
     public void registerModel(final M2ModelResource modelResource) {
-        AuthenticationUtil.runAsSystem(new AuthenticationUtil.RunAsWork<Object>() {
-            @Override
-            public Object doWork() throws Exception {
-                try {
-                    transactionService.getRetryingTransactionHelper().doInTransaction(new RetryingTransactionHelper.RetryingTransactionCallback<Object>() {
-                        @Override
-                        public Object execute() throws Throwable {
-                            try {
-                                if (resourceHelper.nodeDiffersFromResource(modelResource.getResource(), customModelsRepositoryLocation)) {
-                                    repoAdminService.deployModel(
-                                            modelResource.getResource().getInputStream(),
-                                            modelResource.getResource().getFilename());
-                                    logger.debug("Registered model ${modelResource.name}");
-                                }
-                            }
-                            catch (Exception e){
-                                logger.error("Failed to deploy M2Model ${modelResource.name} as a cm:dictionaryModel", e);
-                            }
-                            return null;
+        AuthenticationUtil.runAsSystem(() -> {
+            try {
+                transactionService.getRetryingTransactionHelper().doInTransaction(() -> {
+                    try {
+                        if (resourceHelper.nodeDiffersFromResource(modelResource.getResource(), customModelsRepositoryLocation)) {
+                            repoAdminService.deployModel(
+                                    modelResource.getResource().getInputStream(),
+                                    modelResource.getResource().getFilename());
+                            logger.debug("Registered model {}", modelResource.getName());
                         }
-                    },false,false);
-                    } catch(Exception e){
-                        logger.error("tx error", e);
+                    }
+                    catch (Exception e){
+                        logger.error("Failed to deploy M2Model {} as a cm:dictionaryModel",
+                                modelResource.getName(), e);
                     }
                     return null;
+                },false,false);
+                } catch(Exception e){
+                    logger.error("tx error", e);
                 }
+                return null;
             });
         }
     }

--- a/integration-tests/src/test/java/eu/xenit/dynamicextensionsalfresco/BehaviourTest.java
+++ b/integration-tests/src/test/java/eu/xenit/dynamicextensionsalfresco/BehaviourTest.java
@@ -2,6 +2,7 @@ package eu.xenit.dynamicextensionsalfresco;
 
 import static io.restassured.RestAssured.given;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.isEmptyString;
 
 import org.junit.Test;
 import org.slf4j.Logger;
@@ -17,7 +18,8 @@ public class BehaviourTest extends RestAssuredTest {
 
     @Test
     public void testBehaviour_OnCreateNodePolicy() {
-        logger.info("Test scenario: adding a property to a node using an annotated 'OnCreateNodePolicy' behaviour");
+        logger.info("Test scenario: check that the 'OnCreateNodePolicy' behaviour is triggered when creating an "
+                + "applicable node");
 
         given()
                 .log().ifValidationFails()
@@ -27,5 +29,20 @@ public class BehaviourTest extends RestAssuredTest {
                 .log().ifValidationFails()
                 .statusCode(200)
                 .body(equalTo("\"TestBehaviour successfully processed\""));
+    }
+
+    @Test
+    public void testBehaviour_OnCreateNodePolicy_notTriggeredIfTypeNotApplicable() {
+        logger.info("Test scenario: check that the 'OnCreateNodePolicy' behaviour is not triggered when creating a "
+                + "node that is not applicable");
+
+        given()
+                .log().ifValidationFails()
+                .when()
+                .post("s/dynamic-extensions/testing/behaviours/OnCreateNodePolicyNotTriggeredIfTypeNotApplicable")
+                .then()
+                .log().ifValidationFails()
+                .statusCode(200)
+                .body(isEmptyString());
     }
 }

--- a/integration-tests/test-bundle/src/main/java/eu/xenit/de/testing/behaviours/TestBehaviour.java
+++ b/integration-tests/test-bundle/src/main/java/eu/xenit/de/testing/behaviours/TestBehaviour.java
@@ -12,7 +12,7 @@ import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
 @Component
-@Behaviour("test:document")
+@Behaviour
 @SuppressWarnings("unused")
 public class TestBehaviour implements OnCreateNodePolicy {
 
@@ -22,7 +22,7 @@ public class TestBehaviour implements OnCreateNodePolicy {
     private NodeService nodeService;
 
     @Override
-    @ClassPolicy
+    @ClassPolicy("test:document")
     public void onCreateNode(ChildAssociationRef childAssocRef) {
 
         nodeService.setProperty(

--- a/integration-tests/test-bundle/src/main/java/eu/xenit/de/testing/behaviours/TestBehaviourWebScript.java
+++ b/integration-tests/test-bundle/src/main/java/eu/xenit/de/testing/behaviours/TestBehaviourWebScript.java
@@ -4,7 +4,6 @@ import static eu.xenit.de.testing.Constants.TEST_WEBSCRIPTS_BASE_URI;
 import static eu.xenit.de.testing.Constants.TEST_WEBSCRIPTS_FAMILY;
 
 import com.github.dynamicextensionsalfresco.webscripts.annotations.HttpMethod;
-import com.github.dynamicextensionsalfresco.webscripts.annotations.Transaction;
 import com.github.dynamicextensionsalfresco.webscripts.annotations.Uri;
 import com.github.dynamicextensionsalfresco.webscripts.annotations.WebScript;
 import eu.xenit.de.testing.Model;
@@ -48,13 +47,26 @@ public class TestBehaviourWebScript {
 
     @Uri(value = "/OnCreateNodePolicy", method = HttpMethod.POST)
     public ResponseEntity<String> testBehaviour_OnCreateNodePolicy() {
+        return createTestNodeAndReturnTestBehaviourProperty(
+                Model.TYPE_TESTDOCUMENT,
+                "testBehaviour_OnCreateNodePolicy");
+    }
 
+
+    @Uri(value = "/OnCreateNodePolicyNotTriggeredIfTypeNotApplicable", method = HttpMethod.POST)
+    public ResponseEntity<String> testBehaviour_OnCreateNodePolicy_notTriggeredIfTypeNotApplicable() {
+        return createTestNodeAndReturnTestBehaviourProperty(
+                ContentModel.TYPE_CONTENT,
+                "testBehaviour_OnCreateNodePolicy_notTriggeredIfTypeNotApplicable");
+    }
+
+    private ResponseEntity<String> createTestNodeAndReturnTestBehaviourProperty(final QName type, final String name) {
         final NodeRef createdNode = retryingTransactionHelper.doInTransaction(
                 () -> nodeService.createNode(
                         testFolder,
                         ContentModel.ASSOC_CONTAINS,
-                        QName.createQName("{test.model}", "testBehaviour_OnCreateNodePolicy"),
-                        Model.TYPE_TESTDOCUMENT),
+                        QName.createQName("{test.model}", name),
+                        type),
                 false,
                 true).getChildRef();
 
@@ -66,7 +78,6 @@ public class TestBehaviourWebScript {
                 .getProperty(createdNode, Model.PROP_TESTBEHAVIOURPROPERTY);
 
         return new ResponseEntity<>(testBehaviourProperty, HttpStatus.OK);
-
     }
 
     private NodeRef createOrResetTestFolder() {

--- a/integration-tests/test-bundle/src/main/resources/META-INF/alfresco/models/another-test-model.xml
+++ b/integration-tests/test-bundle/src/main/resources/META-INF/alfresco/models/another-test-model.xml
@@ -1,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<model name="another:anotherTestModel" xmlns="http://www.alfresco.org/model/dictionary/1.0">
+
+    <description>Metadata model for testing purposes, this model depends on the test-model</description>
+    <author>Daan Kerkhofs</author>
+    <version>1.0</version>
+
+    <imports>
+        <import uri="test.model" prefix="test"/>
+    </imports>
+
+    <namespaces>
+        <namespace uri="another.test.model" prefix="another"/>
+    </namespaces>
+
+    <types>
+        <type name="another:document">
+            <title>Another document type</title>
+            <description>Type is child of test:document</description>
+            <parent>test:document</parent>
+        </type>
+    </types>
+
+</model>

